### PR TITLE
Use SearchValues in PathString

### PR DESCRIPTION
**Pull request by MihaZupan**
- Replace `PathStringHelper.IsValidPathChar` loops with `SearchValues`
- Add a vectorized fast-path for longer sections of valid chars
- Use existing `Uri.IsHexEncoding` helper instead of `IsPercentEncodedChar`
- Remove some dead code and eliminate some redundant length checks

Contributes to #46484

I'll post some benchmark numbers later today, but you can easily expect a 10x improvement for long already-valid values.

### CODE REVIEW

**Comment by MihaZupan** [comment]
[Benchmark code](https://gist.github.com/MihaZupan/0b5561bcc4e40e054fd6f9be58b077b8)

| Method |                 Path |       Mean |     Error |    StdDev | Ratio |
|------- |--------------------- |-----------:|----------:|----------:|------:|
| Before |         LongTestPath | 283.907 ns | 2.8201 ns | 4.2210 ns | 16.37 |
|  After |         LongTestPath |  17.331 ns | 0.0495 ns | 0.0693 ns |  1.00 |
|        |                      |            |           |           |       |
| Before | LongT(...)rcent [24] | 302.893 ns | 0.7137 ns | 1.0683 ns | 11.90 |
|  After | LongT(...)rcent [24] |  25.450 ns | 0.0900 ns | 0.1291 ns |  1.00 |
|        |                      |            |           |           |       |
| Before |    LongValidTestPath | 244.094 ns | 1.3209 ns | 1.9771 ns | 32.45 |
|  After |    LongValidTestPath |   7.523 ns | 0.0330 ns | 0.0483 ns |  1.00 |
|        |                      |            |           |           |       |
| Before |             TestPath |  24.910 ns | 0.2281 ns | 0.3271 ns |  1.83 |
|  After |             TestPath |  13.643 ns | 0.0851 ns | 0.1193 ns |  1.00 |

File: src/Http/Http.Abstractions/src/PathString.cs [hunk 1]
```diff
@@ -99,10 +94,15 @@ private static string ToEscapedUriComponent(string value, int i)
         var count = i;
         var requiresEscaping = false;
 
-        while (i < value.Length)
+        while (true)
         {
-            var isPercentEncodedChar = PathStringHelper.IsPercentEncodedChar(value, i);
-            if (PathStringHelper.IsValidPathChar(value[i]) || isPercentEncodedChar)
+            if ((uint)i >= (uint)value.Length)
```
Comment (Tratcher) [inline src/Http/Http.Abstractions/src/PathString.cs:99]: Why not `while ((uint)i < (uint)value.Length)`?

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
components-e2e is being addressed: https://github.com/dotnet/aspnetcore/pull/49113

File: src/Http/Http.Abstractions/src/PathString.cs (hunk 1 above)
Comment (halter73) [reply src/Http/Http.Abstractions/src/PathString.cs:99]: Or even the original `while (i < value.Length)`? Is the cast buying us anything?

**Review by halter73 — APPROVED** [review]
🚀

File: src/Http/Http.Abstractions/src/PathString.cs (hunk 1 above)
Comment (MihaZupan) [reply src/Http/Http.Abstractions/src/PathString.cs:99]: The current pattern is something that the JIT is able to recognize and elide the range check on the `value[i]` access below.
`while ((uint)i < (uint)value.Length)` is something that should [eventually](https://github.com/dotnet/runtime/issues/83349) get you the same result.

Comment (Tratcher) [reply src/Http/Http.Abstractions/src/PathString.cs:99]: Stick with the expected code, let the JIT take care of its own issues.

File: src/Http/Http.Abstractions/src/PathString.cs [hunk 2]
```diff
@@ -99,10 +94,15 @@ private static string ToEscapedUriComponent(string value, int i)
         var count = i;
         var requiresEscaping = false;
 
-        while (i < value.Length)
+        while (true)
```
Comment (MihaZupan) [inline src/Http/Http.Abstractions/src/PathString.cs:97]:
```suggestion
        while ((uint)i < (uint)value.Length)
```

File: src/Http/Http.Abstractions/src/PathString.cs [hunk 3]
```diff
@@ -99,10 +94,15 @@ private static string ToEscapedUriComponent(string value, int i)
...
         {
-            var isPercentEncodedChar = PathStringHelper.IsPercentEncodedChar(value, i);
-            if (PathStringHelper.IsValidPathChar(value[i]) || isPercentEncodedChar)
+            if ((uint)i >= (uint)value.Length)
+            {
+                break;
+            }
+
+            var isPercentEncodedChar = false;
```
Comment (MihaZupan) [inline src/Http/Http.Abstractions/src/PathString.cs:104]:
```suggestion
            var isPercentEncodedChar = false;
```

**Comment by mitchdenny** [comment]
Also related: https://github.com/dotnet/aspnetcore/pull/49114

**Review by Tratcher — APPROVED** [review]

**Pull request merged**